src/AMAMM.sol

```diff
@@ -2,69 +2,200 @@
+        } else {
+            if (targetEpoch - _lastUpdatedEpoch[id] <= K(id)) {
+                return poolEpochManager[id][_lastUpdatedEpoch[id]];
+            } else {
+                return poolEpochManager[id][targetEpoch];
+            }
+        }
+    }
+
+    /// -----------------------------------------------------------------------
+    /// Bidder actions
+    /// -----------------------------------------------------------------------
+    /// @inheritdoc IAmAmm
+    function bid(PoolId id, bytes7 payload, uint128 rent, uint40 _epoch) external virtual override {
+        /// -----------------------------------------------------------------------
+        /// Validation
+        /// -----------------------------------------------------------------------
+
+        address msgSender = LibMulticaller.senderOrSigner();
+        if (_epoch > _getEpoch(id, block.timestamp) + K(id)) {
+            revert AmAmm__BidOutOfBounds();
+        }
+
+        uint128 deposit = depositToken(id, msgSender, rent);
+
+        if (
+            _epoch <= _getEpoch(id, block.timestamp)
+                || rent <= poolEpochManager[id][_epoch].rent.mulWad(MIN_BID_MULTIPLIER(id)) || deposit < rent * K(id)
+                || !_payloadIsValid(id, payload)
+        ) {
+            revert AmAmm__InvalidBid();
+        }
+        Bid memory prevWinner = getLastManager(id, _epoch);
+
+        if (prevWinner.rent != 0) {
+            if (prevWinner.rent < rent) {
+                //Userp Top Bidder and only allow bidder to own N epochs instead of K epochs (N < K)
+                _userBalance[prevWinner.bidder] += _getRefund(id, _lastUpdatedEpoch[id], _epoch); //Refund losing bidder
+                poolEpochManager[id][_epoch] = Bid({bidder: msgSender, payload: payload, rent: rent});
+
+                _userBalance[msgSender] -= _getDeposit(id, _epoch);
+            }
+        } else {
+            poolEpochManager[id][_epoch] = Bid({bidder: msgSender, payload: payload, rent: rent});
+
+            _userBalance[msgSender] -= _getDeposit(id, _epoch);
+        }
+        _updateLastUpdatedEpoch(id, _epoch);
+    }
+
+    function depositToken(PoolId id, address depositor, uint128 rent) internal returns (uint128) {
+        uint128 amount = uint128(rent * K(id));
+
+        if (_userBalance[depositor] >= amount) {
+            uint128 remainderAmount = uint128(_userBalance[depositor] - amount);
+            _pullBidToken(id, depositor, remainderAmount);
+            _userBalance[depositor] += remainderAmount;
+        } else {
+            _pullBidToken(id, depositor, amount);
+            _userBalance[depositor] += amount;
+        }
+
+        return amount;
+    }
+    /// @inheritdoc IAmAmm
+
+    function withdrawBalance(PoolId id, uint128 _amount) external virtual override returns (uint128) {
+        address msgSender = LibMulticaller.senderOrSigner();
+
+        if (!_amAmmEnabled(id) || _userBalance[msgSender] < _amount) {
```

Comment:
Enhancement:
You might need to add `!_amAmmEnabled(id)` as a modifier and apply it to deposit and bid as well